server: TestDrain failed

[(server).TestDrain failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2447026&tab=buildLog) on [master@5403e542f095bd93f5c805a3ac73f3ebe9c82b4b](https://github.com/cockroachdb/cockroach/commits/5403e542f095bd93f5c805a3ac73f3ebe9c82b4b):

```
=== RUN   TestDrain
    test_log_scope.go:155: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDrain318225734
    test_log_scope.go:64: use -show-logs to present logs inline
    stopper.go:98: leaked stopper, created at:
        goroutine 320243 [running]:
        runtime/debug.Stack(0xc004b2b3b8, 0x428998, 0x30)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/util/stop.register(0xc0051fabd0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:52 +0x45
        github.com/cockroachdb/cockroach/pkg/util/stop.NewStopper(0x0, 0x0, 0x0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0x1ca
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).AddServer(0xc004b12900, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:371 +0xbe
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.NewTestCluster(0x5238620, 0xc0009b2000, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:214 +0x5d8
        github.com/cockroachdb/cockroach/pkg/testutils/testcluster.StartTestCluster(0x5238620, 0xc0009b2000, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:130 +0x98
        github.com/cockroachdb/cockroach/pkg/server_test.newTestDrainContext(0xc0009b2000, 0x26)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/drain_test.go:95 +0x85
        github.com/cockroachdb/cockroach/pkg/server_test.doTestDrain(0xc0009b2000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/drain_test.go:40 +0x50
        github.com/cockroachdb/cockroach/pkg/server_test.TestDrain(0xc0009b2000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/server/drain_test.go:35 +0xb5
        testing.tRunner(0xc0009b2000, 0x4ae0c48)
        	/usr/local/go/src/testing/testing.go:1123 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x2b3
--- FAIL: TestDrain (3.53s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDrain PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDrain.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -108,6 +108,11 @@ func (tc *TestCluster) stopServers(ctx context.Context) {
 	for i := range tc.mu.serverStoppers {
 		if tc.mu.serverStoppers[i] != nil {
 			tc.mu.serverStoppers[i].Stop(context.TODO())
+			// We need to check IsStopped() in addition to just waiting for
+			// Stop() above to complete, because if there was an asynchronous
+			// Drain request ongoing, Stop() would take a shortcut and not
+			// actually wait for the async stop to complete.
+			<-tc.mu.serverStoppers[i].IsStopped()
 			tc.mu.serverStoppers[i] = nil
 		}
 	}
@@ -119,6 +124,11 @@ func (tc *TestCluster) StopServer(idx int) {
 	defer tc.mu.Unlock()
 	if tc.mu.serverStoppers[idx] != nil {
 		tc.mu.serverStoppers[idx].Stop(context.TODO())
+		// We need to check IsStopped() in addition to just waiting for
+		// Stop() above to complete, because if there was an asynchronous
+		// Drain request ongoing, Stop() would take a shortcut and not
+		// actually wait for the async stop to complete.
+		<-tc.mu.serverStoppers[idx].IsStopped()
 		tc.mu.serverStoppers[idx] = nil
 	}
 }
@@ -198,6 +208,7 @@ func NewTestCluster(t testing.TB, nodes int, clusterArgs base.TestClusterArgs) *
 				// The Server becomes responsible for closing this.
 				listener, err := net.Listen("tcp", "127.0.0.1:0")
 				if err != nil {
+					tc.Stopper().Stop(context.Background())
 					t.Fatal(err)
 				}
 				firstListener = listener
@@ -212,6 +223,7 @@ func NewTestCluster(t testing.TB, nodes int, clusterArgs base.TestClusterArgs) *
 		}
 
 		if _, err := tc.AddServer(serverArgs); err != nil {
+			tc.Stopper().Stop(context.Background())
 			t.Fatal(err)
 		}
 	}
@@ -242,10 +254,10 @@ func (tc *TestCluster) Start(t testing.TB) {
 
 		if tc.clusterArgs.ParallelStart {
 			go func(i int) {
-				errCh <- tc.StartServer(tc.Server(i), tc.serverArgs[i])
+				errCh <- tc.startServer(i, tc.serverArgs[i])
 			}(i)
 		} else {
-			if err := tc.StartServer(tc.Server(i), tc.serverArgs[i]); err != nil {
+			if err := tc.startServer(i, tc.serverArgs[i]); err != nil {
 				t.Fatal(err)
 			}
 			// We want to wait for stores for each server in order to have predictable
@@ -340,12 +352,12 @@ func (tc *TestCluster) AddAndStartServer(t testing.TB, serverArgs base.TestServe
 	if serverArgs.JoinAddr == "" && len(tc.Servers) > 0 {
 		serverArgs.JoinAddr = tc.Servers[0].ServingRPCAddr()
 	}
-	serv, err := tc.AddServer(serverArgs)
+	_, err := tc.AddServer(serverArgs)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	if err := tc.StartServer(serv, serverArgs); err != nil {
+	if err := tc.startServer(len(tc.Servers)-1, serverArgs); err != nil {
 		t.Fatal(err)
 	}
 }
@@ -368,7 +380,6 @@ func (tc *TestCluster) AddServer(serverArgs base.TestServerArgs) (*server.TestSe
 	); err != nil {
 		return nil, err
 	}
-	serverArgs.Stopper = stop.NewStopper()
 	if tc.clusterArgs.ReplicationMode == base.ReplicationManual {
 		var stkCopy kvserver.StoreTestingKnobs
 		if stk := serverArgs.Knobs.Store; stk != nil {
@@ -401,14 +412,18 @@ func (tc *TestCluster) AddServer(serverArgs base.TestServerArgs) (*server.TestSe
 
 	tc.Servers = append(tc.Servers, s)
 	tc.serverArgs = append(tc.serverArgs, serverArgs)
+
+	tc.mu.Lock()
+	defer tc.mu.Unlock()
+	thisStopper := s.Stopper()
+	tc.mu.serverStoppers = append(tc.mu.serverStoppers, thisStopper)
 	return s, nil
 }
 
-// StartServer is the companion method to AddServer, and is responsible for
+// startServer is the companion method to AddServer, and is responsible for
 // actually starting the server.
-func (tc *TestCluster) StartServer(
-	server serverutils.TestServerInterface, serverArgs base.TestServerArgs,
-) error {
+func (tc *TestCluster) startServer(idx int, serverArgs base.TestServerArgs) error {
+	server := tc.Servers[idx]
 	if err := server.Start(); err != nil {
 		return err
 	}
@@ -418,11 +433,7 @@ func (tc *TestCluster) StartServer(
 		return err
 	}
 
-	tc.mu.Lock()
-	defer tc.mu.Unlock()
-
 	tc.Conns = append(tc.Conns, dbConn)
-	tc.mu.serverStoppers = append(tc.mu.serverStoppers, server.Stopper())
 	return nil
 }
 

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -108,11 +108,6 @@ func (tc *TestCluster) stopServers(ctx context.Context) {
 	for i := range tc.mu.serverStoppers {
 		if tc.mu.serverStoppers[i] != nil {
 			tc.mu.serverStoppers[i].Stop(context.TODO())
-			// We need to check IsStopped() in addition to just waiting for
-			// Stop() above to complete, because if there was an asynchronous
-			// Drain request ongoing, Stop() would take a shortcut and not
-			// actually wait for the async stop to complete.
-			<-tc.mu.serverStoppers[i].IsStopped()
 			tc.mu.serverStoppers[i] = nil
 		}
 	}
@@ -124,11 +119,6 @@ func (tc *TestCluster) StopServer(idx int) {
 	defer tc.mu.Unlock()
 	if tc.mu.serverStoppers[idx] != nil {
 		tc.mu.serverStoppers[idx].Stop(context.TODO())
-		// We need to check IsStopped() in addition to just waiting for
-		// Stop() above to complete, because if there was an asynchronous
-		// Drain request ongoing, Stop() would take a shortcut and not
-		// actually wait for the async stop to complete.
-		<-tc.mu.serverStoppers[idx].IsStopped()
 		tc.mu.serverStoppers[idx] = nil
 	}
 }

--- a/pkg/util/stop/stopper.go
+++ b/pkg/util/stop/stopper.go
@@ -475,6 +475,8 @@ func (s *Stopper) Stop(ctx context.Context) {
 	s.mu.Unlock()
 
 	if stopCalled {
+		// Wait for the concurrent Stop() to complete.
+		<-s.stopped
 		return
 	}
 

